hyperstart/api/grpc/oci.proto

```diff
@@ -0,0 +1,456 @@
+	string dummy = 1;
+}
+
+message LinuxIDMapping {
+	// HostID is the starting UID/GID on the host to be mapped to 'ContainerID'
+	uint32 HostID = 1;
+
+	// ContainerID is the starting UID/GID in the container
+	uint32 ContainerID = 2;
+
+	// Size is the number of IDs to be mapped
+	uint32 Size = 3;
+}
+
+message LinuxNamespace {
+	// Type is the type of namespace
+	string Type = 1;
+
+	// Path is a path to an existing namespace persisted on disk that can be joined
+	// and is of the same type
+	string Path = 2;
+}
+
+message LinuxDevice {
+	// Path to the device.
+	string Path = 1;
+
+	// Device type, block, char, etc.
+	string Type = 2;
+
+	// Major is the device's major number.
+	int64 Major = 3;
+
+	// Minor is the device's minor number.
+	int64 Minor = 4;
+
+	// FileMode permission bits for the device.
+	uint32 FileMode = 5;
+
+	// UID of the device.
+	uint32 UID = 6;
+
+	// Gid of the device.
+	uint32 GID = 7;
+}
+
+message LinuxResources {
+	// Devices configures the device whitelist.
+	repeated LinuxDeviceCgroup Devices = 1;
+
+	// Memory restriction configuration
+	LinuxMemory Memory = 2;
+
+	// CPU resource restriction configuration
+	LinuxCPU CPU = 3;
+
+	// Task resource restriction configuration.
+	LinuxPids Pids = 4;
+
+	// BlockIO restriction configuration
+	LinuxBlockIO BlockIO = 5;
+
+	// Hugetlb limit (in bytes)
+	repeated LinuxHugepageLimit HugepageLimits = 6;
+
+	// Network restriction configuration
+	LinuxNetwork Network = 7;
+}
+
+message LinuxMemory {
+	// Memory limit (in bytes).
+	int64 Limit = 1;
+
+	// Memory reservation or soft_limit (in bytes).
+	int64 Reservation = 2;
+
+	// Total memory limit (memory + swap).
+	int64 Swap = 3;
+
+	// Kernel memory limit (in bytes).
+	int64 Kernel = 4;
+
+	// Kernel memory limit for tcp (in bytes)
+	int64 KernelTCP = 5;
+
+	// How aggressive the kernel will swap memory pages.
+	uint64 Swappiness = 6;
+
+	// DisableOOMKiller disables the OOM killer for out of memory conditions
+	bool DisableOOMKiller = 7;
+}
+
+message LinuxCPU {
+	// CPU shares (relative weight (ratio) vs. other cgroups with cpu shares).
+	uint64 Shares = 1;
+
+	// CPU hardcap limit (in usecs). Allowed cpu time in a given period.
+	int64 Quota = 2;
+
+	// CPU period to be used for hardcapping (in usecs).
+	uint64 Period = 3;
+
+	// How much time realtime scheduling may use (in usecs).
+	int64 RealtimeRuntime = 4;
+
+	// CPU period to be used for realtime scheduling (in usecs).
+	uint64 RealtimePeriod = 5;
+
+	// CPUs to use within the cpuset. Default is to use any CPU available.
+	string Cpus = 6;
+
+	// List of memory nodes in the cpuset. Default is to use any available memory node.
+	string Mems = 7;
+}
+
+message linuxBlockIODevice {
+	// Major is the device's major number.
+	int64 Major = 1;
+
+	// Minor is the device's minor number.
+	int64 Minor = 2;
+}
+
+message LinuxWeightDevice {
+	// Major is the device's major number.
+	int64 Major = 1;
+
+	// Minor is the device's minor number.
+	int64 Minor = 2;
```

Comment:
The test doesn't test the copying about it, which is required and important since LinuxWeightDevice in oci.Specs embeds linuxBlockIODevice as an anonymous field here.

Reply:
Good point. I fixed it now and the code supports copying embedded structures.